### ISSUE TITLE
Expected an assignment or function call and instead saw an expression

### ISSUE THREAD

**Issue by asaadawey** [typescript-eslint/typescript-eslint#2866]
- [T] I have tried restarting my IDE and the issue persists.
- [T ] I have updated to the latest version of the packages.
- [T ] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

```JSON
    {
      "root": true,
      "parser": "@typescript-eslint/parser",
      "plugins": ["@typescript-eslint"],
      "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/eslint-recommended",
        "plugin:@typescript-eslint/recommended",
        "prettier/@typescript-eslint", // Uses eslint-config-prettier to disable ESLint rules from @typescript-eslint/eslint-plugin that would conflict with prettier
        "plugin:prettier/recommended"
      ],
      "overrides": [
        {
          "files": ["*.tsx", "*.ts"],
          "rules": {
            "no-unused-expressions": "off",
            "@typescript-eslint/no-unused-expressions": ["off"]
          }
        }
      ],
      "rules": {
        "no-ternary": "off",
        "no-unused-expressions": "off",
        "@typescript-eslint/no-unused-expressions": ["off"]
      }
    }
```

```TS
      .then(({ data }: FetchAllVehiclesQueryHookResult) => { //135
        data?.vehicles.forEach((item) => { ..... //136
```

**Expected Result**

**Actual Result**

**Additional Info**

**Versions**

| package                                   | version |
| ----------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin-tslint` | `4.9.1 |
| `@typescript-eslint/parser`               | `4.9.1` |
| `TypeScript`                              | `4.1.2` |
| `ESLint`                                  | `7.15.0` |
| `node`                                    | `X.Y.Z` |

**Comment by bradzacher** [comment]
You've skipped a lot of sections in the template.

The issue template exists so that you provide all the necessary information up front to help with investigation

Could you provide an isolated repro?

Also what is the error you're seeing?

**Comment by KhodeN** [comment]
The error "Expected an assignment or function call and instead saw an expression  @typescript-eslint/no-unused-expressions" appears for expressions like 

```onCancel?.();```

**Comment by bradzacher** [comment]
Going to need more information. Please file a new issue.
